core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.linecorp.armeria.client.endpoint.healthcheck.HealthCheckedEndpointGroup.DEFAULT_HEALTH_CHECK_RETRY_BACKOFF;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.function.Function;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ClientOptionsBuilder;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.AsyncCloseable;
+
+/**
+ * A skeletal builder implementation for creating a new {@link HealthCheckedEndpointGroup}.
+ */
+public abstract class AbstractHealthCheckedEndpointGroupBuilder {
```

Comment:
That's a good question. If we use that approach, the generated Javadoc will look ugly.

![image](https://user-images.githubusercontent.com/173918/62597394-ded36a00-b920-11e9-88b6-a510ae5b2b4e.png)

Note the return type `B`.

The self-type parameters in `AbstractRequestContextBuilder` and some other builders are the crufts from the past.

Reply:
I didn't know regarding Javadoc result. I learned a good tip. 💯

Reply:
> Could be package-private?

No, it is public intentionally, so that a user can extend it and send custom health check requests, such as gRPC calls.

Reply:
Ah, thanks! I thought we have
```
HealthCheckedEndpointGroupBuilder(
    EndpointGroup delegate, Function<? super HealthCheckerContext, ? extends AsyncCloseable> checker) {
```
so that a user can customize the health check request.
The constructor looks ugly though. 😆